### ISSUE TITLE
Array init problem

### ISSUE THREAD

**Issue by tamlin-mike** [icsharpcode/ILSpy#1336]
The following compiled with csc 2.10.0.0
`class C {
	void fn() { var a = new int[200];a[0]=1;a[1]=2; }
}
`
Generates... a lot of decompiled lines:

```
int[] array = new int[200]
{
  1,
  2,
  0,
[...]
```
Imagine replacing the array size with 2^20.

I'm not sure what heuristic cutoff one should aim for here, and if it should only be based on number of non-zero-initialized elements or if it should take a percentage zero/non-zero into account, but for a starter it could probably suffice to revert such cases back to ILSpy's previous behavior.

**Comment by siegfriedpammer** [comment]
would

```
if (elementCount < length / 3)
    return false;
```

be a good cutoff condition?

ILSpy 2.4 did something like this:

Following condition would be placed inside the for loop of `HandleSimpleArrayInitializer`:
```
if (index - nextMinimumIndex > 300)
    return false;
```

However, this would not solve the problem discribed in this issue, it would just revive the old cutoff behavior.

Maybe applying both changes is best?

**Comment by tamlin-mike** [comment]
While I haven't looked at the code yet, I have tried to formulate some kind of description of potential heuristics.

Going for a third could be OK for smaller arrays, but would still not catch the problem with larger ones as you also noted.

For not-default initialized (say a CRC32 table) it's obviously a no-brainer, but let's say someone lifted some C++ code into IL-land, let's say it looks like `static int foo[512] = { 1,2,3 }; foo[511] = 47;` (`foo[3] .. foo[510]` would be zero).

For (even large) arrays where (almost) all elements are initialized to a non-default value, I think it would be correct to place them all as initializers. For that case (or maybe aways?) perhaps the cutoff could be log2(length) relative, since array-sizes can obviously vary wildly.

To make it more fun, this is also a recursive problem, if initializing multi-dimensional arrays (something ILSpy currently struggles with, and fails, at least when the elements are non-POD). Let's say we have `int[4][17]`, where the second inner array of 17 is `{ 1,2,0,0... }` but the rest are initialized with non-zero values. Do we want that single instance (that at `n / 3` cutoff would trigger) to abort all initialization? For such a scenario, probably not.

It's an interesting problem. I'm not sure how to cover all bases (if it's even possible), but hopefully I provided some food for thought.

**Comment by siegfriedpammer** [comment]
> For (even large) arrays where (almost) all elements are initialized to a non-default value, I think it would be correct to place them all as initializers. For that case (or maybe aways?) perhaps the cutoff could be log2(length) relative, since array-sizes can obviously vary wildly.

In the case of the example I mentioned above, this would be `log2(65536) =16`, which would not trigger, because there are more than 16 non-default values in the list. Or, maybe I am misunderstanding how the cut-off would be implemented in case of log2.

**Comment by tamlin-mike** [comment]
> what do you mean by "POD"?

Term from C/C++: "Plain Old Data".

> 65536 items (most of them null) is a bit too much.

Hahaha. Perhaps a tad excessive, yes. :-)

> Implementing the "a third" rule would solve this.

That should be enough evidence to just do it.

> Or, maybe I am misunderstanding how the cut-off would be implemented in case of log2.

I was a bit unclear, and probably hadn't thought it through properly. IIRC I was probably thinking of comparing log2(array size) vs. log2(# of initialized elements). Not sure what I thought that would accomplish. Let's forget I mentioned it.

I'll try to find a small-ish example of 2-dimensional array init failing, where ILSpy instead reverts to manual assignment of each array element, and follow up here when I do.

**Comment by tamlin-mike** [comment]
Example of failing to array-init. Test-compiled with csc 2.10.0.0, followed by letting ILSpy do its thing on the generated dll.
```
public class C {
	class V3f { float x,y,z; public V3f(float _x, float _y, float _z) { x=_x;y=_y;z=_z; } }
	// decompilation of rg1 init fails. It becomes 9 separate (local) objects, each
	// manually assigned to rg1[x,y]
	static V3f[,] rg1 = new V3f[3,3] {
		{ new V3f(1,1,1), new V3f(1,1,1), new V3f(1,1,1) },
		{ new V3f(1,1,1), new V3f(1,1,1), new V3f(1,1,1) },
		{ new V3f(1,1,1), new V3f(1,1,1), new V3f(1,1,1) }
	};
	// rg2 "works". ILSpy still puts the init code in "manual" static ctor, instead of field init.
	static int[,] rg2 = new int[3,3] {
		{ 1,1,1 },
		{ 1,1,1 },
		{ 1,1,1 }
	};
}
```

**Comment by dgrunwald** [comment]
We definitely need some limit to avoid emitting tons of default labels.
I think a fixed fraction like `if (explicitElementCount < length / 3 + 5)` should be fine; this way the C# code size will still only grow linearly with the IL code size. (the addition of a small constant is necessary so that very small arrays ending with a default element still get to use array initializer syntax)

The performance issue with long initializer lists has a separate ticket: #1202

**Comment by siegfriedpammer** [comment]
IL Ast of the rg1 initializer (used different numbers for clarity):

```
[0] stloc S_2(newarr ConsoleApp2.C+V3f(ldc.i4 3, ldc.i4 3))
[1] stloc S_9(newobj V3f..ctor(ldc.f4 1, ldc.f4 1, ldc.f4 1))
[2] stobj ConsoleApp2.C+V3f(ldelema ConsoleApp2.C+V3f(ldloc S_2, ldc.i4 0, ldc.i4 0), ldloc S_9)
[3] stloc S_16(newobj V3f..ctor(ldc.f4 2, ldc.f4 2, ldc.f4 2))
[4] stobj ConsoleApp2.C+V3f(ldelema ConsoleApp2.C+V3f(ldloc S_2, ldc.i4 0, ldc.i4 1), ldloc S_16)
[5] stloc S_23(newobj V3f..ctor(ldc.f4 3, ldc.f4 3, ldc.f4 3))
[6] stobj ConsoleApp2.C+V3f(ldelema ConsoleApp2.C+V3f(ldloc S_2, ldc.i4 0, ldc.i4 2), ldloc S_23)
[7] stloc S_30(newobj V3f..ctor(ldc.f4 2, ldc.f4 2, ldc.f4 2))
[8] stobj ConsoleApp2.C+V3f(ldelema ConsoleApp2.C+V3f(ldloc S_2, ldc.i4 1, ldc.i4 0), ldloc S_30)
[9] stloc S_37(newobj V3f..ctor(ldc.f4 3, ldc.f4 3, ldc.f4 3))
[10] stobj ConsoleApp2.C+V3f(ldelema ConsoleApp2.C+V3f(ldloc S_2, ldc.i4 1, ldc.i4 1), ldloc S_37)
[11] stloc S_44(newobj V3f..ctor(ldc.f4 4, ldc.f4 4, ldc.f4 4))
[12] stobj ConsoleApp2.C+V3f(ldelema ConsoleApp2.C+V3f(ldloc S_2, ldc.i4 1, ldc.i4 2), ldloc S_44)
[13] stloc S_51(newobj V3f..ctor(ldc.f4 3, ldc.f4 3, ldc.f4 3))
[14] stobj ConsoleApp2.C+V3f(ldelema ConsoleApp2.C+V3f(ldloc S_2, ldc.i4 2, ldc.i4 0), ldloc S_51)
[15] stloc S_58(newobj V3f..ctor(ldc.f4 4, ldc.f4 4, ldc.f4 4))
[16] stobj ConsoleApp2.C+V3f(ldelema ConsoleApp2.C+V3f(ldloc S_2, ldc.i4 2, ldc.i4 1), ldloc S_58)
[17] stloc S_65(newobj V3f..ctor(ldc.f4 5, ldc.f4 5, ldc.f4 5))
[18] stobj ConsoleApp2.C+V3f(ldelema ConsoleApp2.C+V3f(ldloc S_2, ldc.i4 2, ldc.i4 2), ldloc S_65)
```

Currently we're unable to build an array initializer for this because `newobj` cannot be moved over `ldelema` due to side-effects. @dgrunwald do we need a special case for this?

**Comment by dgrunwald** [comment]
When we decompose a `stelem` into `ldelema`+`stobj`, we effectively change when the out-of-bounds exception happens (before or after the value-being-stored was evaluated), so we mark the `ldelema` as having a "delayed exception" (this flag might be missing in the ILAst's text representation?).
This is what allows the value expression to be inlined in the single-dimensional array initializer case.

Is it an actual difference in the order of operations in the IL code between single-dimensonal and multi-dimensional initializers? Or are we just forgetting to set "delayed exception" mode for stores to multi-dimensional arrays?

**Comment by siegfriedpammer** [comment]
Thanks for clearing that up! I forgot that we had this flag. I remember discussing this a few years back, when we got started with the newdecompiler branch.

Yes, the `DelayedExceptions` flag, was missing from the ILAst output, and ILReader did not set the flag on Get/Set methods used with arrays. Should I set the flag on instructions inside an array initializer? I think it will not be of much use anymore.
